### PR TITLE
Add batch functionality

### LINKED ISSUE #9889: The `addBatch` SQL Feature Fails in the JDBC When Writing a PySpark DataFrame to a DuckDB Database
### What happens?

When writing a PySpark DataFrame to a DuckDB database, the following exception is reported: `java.sql.SQLFeatureNotSupportedException: addBatch`. Could the DuckDB JDBC support the `addBatch` feature please?

## Reproduction
```
import pyspark.sql as psql
spark = psql.SparkSession.builder.config('spark.jars', 'path/to/duckdb_jdbc-0.8.1.jar').getOrCreate()
df = spark.createDataFrame([
    (1, 2.2, 'string1'),
    (2, 3.3, 'string2'),
    (3, 4.4, 'string3')
], schema='col1 long, col2 double, col3 string')
url = 'jdbc:duckdb:path/to/my-duckdb-database.db'
writer = psql.DataFrameWriter(df)
writer.jdbc(
    table='Test', url=url, mode='append',
    properties={'driver': 'org.duckdb.DuckDBDriver'})
```

## Error message
```
Py4JJavaError: An error occurred while calling o44.jdbc.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 11 in stage 0.0 failed 1 times, most recent failure: Lost task 11.0 in stage 0.0 (TID 11) (dionysus executor driver): java.sql.SQLFeatureNotSupportedException: addBatch
	at org.duckdb.DuckDBPreparedStatement.addBatch(DuckDBPreparedStatement.java:711)
...
```

### OS:

Linux

### DuckDB Version:

0.8.1

### DuckDB Client:

Python

### Full Name:

Erik Huckvale

### Affiliation:

University of Kentucky

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by osidekyle**
Closes #9889 

Updates:
* Add batch write functionality with prepared statements
* Added functionality to addBatch and executeBatch methods, added check to ensure executeBatch is used once batched values are added to the prepared statement
* Add test cases for the above functionality
* Note: not sure if we need the addBatch(String sql) functionality because it appears a DuckDBPreparedStatement object cannot be created without passing in the SQL string to the constructor. Am I wrong on this?

This is my first contribution to duckdb so all advice is greatly welcomed!

### CODE REVIEW

File: tools/jdbc/src/main/java/org/duckdb/DuckDBPreparedStatement.java [hunk 1]
```diff
@@ -445,17 +453,29 @@ public int getResultSetType() throws SQLException {
 
     @Override
     public void addBatch(String sql) throws SQLException {
```
Comment (Mause) [inline tools/jdbc/src/main/java/org/duckdb/DuckDBPreparedStatement.java:455]: Can you add a test for this method/include it in an existing test please?

**Review by Mause — CHANGES_REQUESTED** [review]

File: tools/jdbc/src/main/java/org/duckdb/DuckDBPreparedStatement.java (hunk 1 above)
Comment (osidekyle) [reply tools/jdbc/src/main/java/org/duckdb/DuckDBPreparedStatement.java:455]: Definitely!

**Comment by Mause** [comment]
> Note: not sure if we need the addBatch(String sql) functionality because it appears a DuckDBPreparedStatement object cannot be created without passing in the SQL string to the constructor. Am I wrong on this?

You can call it on an existing instance though, no? Or via `createStatement().addBatch(...)`

**Comment by osidekyle** [comment]
> > Note: not sure if we need the addBatch(String sql) functionality because it appears a DuckDBPreparedStatement object cannot be created without passing in the SQL string to the constructor. Am I wrong on this?
> 
> You can call it on an existing instance though, no? Or via `createStatement().addBatch(...)`

Yup, I started thinking about how to write a test realized it could be called on an existing instance.

**Comment by Mause** [comment]
> > > Note: not sure if we need the addBatch(String sql) functionality because it appears a DuckDBPreparedStatement object cannot be created without passing in the SQL string to the constructor. Am I wrong on this?
> > 
> > You can call it on an existing instance though, no? Or via `createStatement().addBatch(...)`
> 
> Now looking at the Java docs for addBatch(String sql) of Statement: https://docs.oracle.com/javase/8/docs/api/java/sql/Statement.html#addBatch-java.lang.String-
> 
> "Note:This method cannot be called on a PreparedStatement or CallableStatement."
> 
> Maybe we don't need to implement addBatch(String sql) for DuckDBPreparedStatement? It kind of makes sense since we can't add SQL statement after setting the parameter values.

Sure, but we actually use the same class to implement Statement

**Comment by osidekyle** [comment]
> > > > Note: not sure if we need the addBatch(String sql) functionality because it appears a DuckDBPreparedStatement object cannot be created without passing in the SQL string to the constructor. Am I wrong on this?
> > > 
> > > 
> > > You can call it on an existing instance though, no? Or via `createStatement().addBatch(...)`
> > 
> > 
> > Now looking at the Java docs for addBatch(String sql) of Statement: https://docs.oracle.com/javase/8/docs/api/java/sql/Statement.html#addBatch-java.lang.String-
> > "Note:This method cannot be called on a PreparedStatement or CallableStatement."
> > Maybe we don't need to implement addBatch(String sql) for DuckDBPreparedStatement? It kind of makes sense since we can't add SQL statement after setting the parameter values.
> 
> Sure, but we actually use the same class to implement Statement

Ah, so this addBatch(String sql) should be the Statement implementation? Sorry for discussing this in the PR!

**Comment by Mause** [comment]
> > > > > Note: not sure if we need the addBatch(String sql) functionality because it appears a DuckDBPreparedStatement object cannot be created without passing in the SQL string to the constructor. Am I wrong on this?
> > > > 
> > > > 
> > > > You can call it on an existing instance though, no? Or via `createStatement().addBatch(...)`
> > > 
> > > 
> > > Now looking at the Java docs for addBatch(String sql) of Statement: https://docs.oracle.com/javase/8/docs/api/java/sql/Statement.html#addBatch-java.lang.String-
> > > "Note:This method cannot be called on a PreparedStatement or CallableStatement."
> > > Maybe we don't need to implement addBatch(String sql) for DuckDBPreparedStatement? It kind of makes sense since we can't add SQL statement after setting the parameter values.
> > 
> > Sure, but we actually use the same class to implement Statement
> 
> Ah, so this addBatch(String sql) should be the Statement implementation? Sorry for discussing this in the PR!

We use the same class for both actually

Also feel free to reach out on discord if that's easier: https://discord.duckdb.org

**Comment by osidekyle** [comment]
@szarnyasg Sorry for my lack of knowledge, but is this documentation something I need to add? If so, where will I add that?

**Comment by Mytherin** [comment]
@Mause is this ready to merge?

> @szarnyasg Sorry for my lack of knowledge, but is this documentation something I need to add? If so, where will I add that?

The label is meant for adding documentation to the [website](https://github.com/duckdb/duckdb-web), so not required for this PR directly.

**Comment by Mause** [comment]
> @Mause is this ready to merge?
> 
> > @szarnyasg Sorry for my lack of knowledge, but is this documentation something I need to add? If so, where will I add that?
> 
> The label is meant for adding documentation to the [website](https://github.com/duckdb/duckdb-web), so not required for this PR directly.

Looks good to me! We may want to investigate splitting this class down the road though

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**